### DISCUSSION TITLE
Get info for checkboxes?

### DISCUSSION THREAD

**Discussion by Rapid1898-code** [jsvine/pdfplumber#738] [Ask for help with specific PDFs]
Hello, i am able to extract the whole text and also the tables from the pdf what is great -
But is it also possible to get the information for the checkboxes if they are checked or not?
If yes how can i get this information in the attached pdf?

[input.pdf](https://github.com/jsvine/pdfplumber/files/9737476/input.pdf)

**Comment by samkit-jain** [comment]
Hi @Rapid1898-code , appreciate your interest in the library. In the case of the PDF, that you shared, to get all the checkboxes, you can filter out the rect objects by
```python
# Getting all the rect objects that are of square shape and in the range of 10-15 units per side and having no stroking color
checkboxes_all = [rect for rect in rects if 10 < rect["width"] < 15 and 10 < rect["height"] < 15 and int(rect["width"]) == int(rect["height"]) and rect['non_stroking_color'] == None]
```
![checkboxes_all](https://user-images.githubusercontent.com/15127115/194819417-0b023ca2-f9ca-4a0a-aa14-029ec5507857.png)

Now, to get all the checkboxes that are filled (aka checked), you can filter out by
```python
checkboxes_checked = []

# If a checkbox has a diagonal line within it, consider it checked.
for checkbox in checkboxes_all:
    # Crop the page to contain just the checkbox rectangle.
    cropped = page.crop((checkbox["x0"], checkbox["top"], checkbox["x1"], checkbox["bottom"]))
    
    # If the cropped page has 6 edges, then consider it checked. Unchecked ones would have just 4 lines.
    if len(cropped.edges) == 6:
        checkboxes_checked.append(checkbox)
```

![checkboxes_checked](https://user-images.githubusercontent.com/15127115/194820983-c764df35-912d-4caa-86c8-980fd166f6f0.png)


You would need to do further post processing to know the text that is besides these checkboxes but I hope the logic would put you in the right direction.

**Comment by Rapid1898-code** [comment]
Hello - thanks a lot for your response -
I think i understand how you do this.
What is the best way to also get the text (beside the checkboxes)?

Should i get first the checkboxes as you described it above.
Then also get all words using extract_words().
And then sort both informations together using the x0/y0 coordinates of this elements.

Or is there some better / easier way to do this?

KR, Max

**Reply by samkit-jain** [reply]
Yes, my recommendation would also be the same. An easier way, at least for the last section would be to extract the data as a table but that won't work for the other sections.

**Comment by Rapid1898-code** [comment]
ok thanks again a lot - may i bother you with a similar question?
[input.pdf](https://github.com/jsvine/pdfplumber/files/9745743/input.pdf)
I think handling checkboxes is more or less clear i think.
But what can i do when i want to match radio buttons?
Find attached a second pdf where you can find some blue radio buttons in it.
Any idea how to check if this radio-buttons are checked or unchecked?

**Reply by samkit-jain** [reply]
For the radio button, I would go for iterating over the curves. To get all the radio buttons, you can do
```python
circles = [curve for curve in page.curves if len(curve["points"]) > 30 and curve["height"] > 12]
```
![circles](https://user-images.githubusercontent.com/15127115/194866420-a83860f9-e7ff-48a6-82b2-7e7fbd7b6882.png)
and then cropping again to find circles within circles
```python
selected_circles = []

for circle in circles:
    cropped = page.crop((circle["x0"], circle["top"], circle["x1"], circle["bottom"]))
    if any([curve["height"] < 12 for curve in cropped.curves]):
        selected_circles.append(circle)
```
![selected](https://user-images.githubusercontent.com/15127115/194868165-d3bc9a2c-5c0e-4bcb-9099-543762d70003.png)

**Comment by Rapid1898-code** [comment]
Thanks a lot for your help again - this works great.
Can you probably also send me the code how to highlight the checkboxe / radiobutton using the visual debugging functionality from pdfplumber?

**Reply by samkit-jain** [reply]
Sure,

```python
im = page.to_image(resolution=200)
im.draw_lines(circles)  # This circles is the same as in my comments above. You can use page.edges, page.curves, ... as well
im.save("/path/to/image.png", format="PNG")
```

**Comment by Rapid1898-code** [comment]
Thanks that worked great for the radio-buttons.
I also tried this with the cb-pdf with the following-code:

```
  rects = page.rects
  checkboxesALL = [rect for rect in rects 
                    if 10 < rect["width"] < 15 
                    and 10 < rect["height"] < 15 
                    and int(rect["width"]) == int(rect["height"]) 
                    and rect['non_stroking_color'] == None]
  im = page.to_image(resolution=200) 
  im.draw_lines(checkboxesALL)
  fn = os.path.join(path, "imageCB.png")
  im.save(fn, format="PNG")
```

It generally works - but as you can see in the attached picture only the diagonal line from left-top to right-bottom is red.
You made the whole rectangle red - how have you achieved this?

![imageCB](https://user-images.githubusercontent.com/57018730/195161369-f3662a60-0662-4e05-a0f0-88b7465d60eb.png)

**Reply by samkit-jain** [reply]
Use `im.draw_rects(checkboxesALL)`

**Comment by Rapid1898-code** [comment]
Thanks - works great!
